### ISSUE TITLE
RPC: Decision tracking issue

### ISSUE THREAD

**Issue by camdencheek** [sourcegraph/sourcegraph#33863]
This is a tracking issue that will capture decisions related to introducing gRPC at Sourcegraph.

Eventually, all of these tasks will be converted to separate issues for discussion.

Status: Currently, @camdencheek is doing some discovery on migrating the searcher API to use gRPC. No decisions are being made during this spike. The goal is to build out a MVP and discover what decisions will need to be made. Those will be recorded here.

Tasks:
- [x] #35351
- [x] #34439
- [x] #35352
- [ ] Establish patterns for code organization
- [x] #35361
- [ ] Establish patterns for [error handling](https://grpc.io/docs/guides/error/)
- [ ] Establish naming conventions (api vs protocol, request/response naming)

**Comment by bobheadxi** [comment]
Linking for context: https://github.com/sourcegraph/sourcegraph/discussions/33283

**Comment by camdencheek** [comment]
**Update on my explorations:**

I implemented gRPC for a subset of the single searcher endpoint. It is available in a rough form on the `cc/grpc-searcher` branch. The tasks above capture the open questions that came up as I was looking at this. They are in rough order of importance.

**Some discussion on the items above:**

The TLS stuff is something we should come to consensus on before anything is merged. Right now, I'm solving it using `insecure.NewCredentials()` on the client side and `h2c.NewHandler`, then multiplexing requests on the `Content-Type` header (gRPC set it to `application/grpc`). I do this to avoid opening a new port to get this working, but I honestly have no idea how questionable it is. The [documentation](https://sourcegraph.com/github.com/golang/net/-/blob/http2/h2c/h2c.go?L55-63&subtree=true) talking about hijacking http/2 requests is a bit...hacky...sounding. 

For toolchain, we are already using [buf](https://github.com/bufbuild/buf) elsewhere in the codebase, but it seems that's only for protobuf and not for gRPC. I did not attempt to use it for this exploration, instead just sticking to vanilla protoc.

For load balancing, it seems like gRPC provides some fairly flexible tools for doing client-side load balancing, so it seems possible to replace the [pretty manual](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@a749e93e95b47c8b3fb0409d0840678c52add810/-/blob/internal/search/searcher/client.go?L93-109) client-side load balancing with something more transparent to the RPC clients. This is probably not necessary for a first pass, but it's a possible nice bonus.
